### ISSUE TITLE
Python Server & Node Client , On Authentication Fail client receives 1 fix error in any scenario.

### ISSUE THREAD

**Issue by harshkoralwala** [miguelgrinberg/python-socketio#338]
Hello,

I have a scenario where HTTPS Server is of Python and Client is from the nodeJs,

For a client, I need to show 2 different messages for the below scenario
     -If a client tries to connect with invalid URL like 'abc.com', I need to show the message "server not found"
    -If the user enters valid URL and passes an invalid token, I need to show the message "Invalid token"

On Authentication fail if i raise any type of errors on server,on client i receive only 1 type of FIX error that is "**websocket error**".I can not receive same error object with text on client sidewhich is sent by server.

For Node, I have just 1 default function where I can receive any type of error on connection error...

```
socket.on("connect_error", (data) => {
    console.log((data));
})
```

So here if from server i return _ConnectionRefusedError('authentication failed')_ (refer server code),so in node js i can compare datatype and text message of function parameter,and on its basis i can decide which message i need  to show to user.
   

> My Python server HTTPS server +SocketIO code looks like:

```
import socketio
from aiohttp import web
import asyncio
import eventlet
import ssl
import jwt
from urllib.parse import urlparse, parse_qs

sio = socketio.AsyncServer(cors_allowed_origins="*")

app = web.Application()
sio.attach(app)

@sio.event
async def connect(sid, environ):
    print('>>>> connect <<<<< ')
    print(sid)
    raise ConnectionRefusedError('authentication failed')

@sio.event
def disconnect(sid):
    print('>>>> disconnect <<<<< ')


@sio.event
def error(sid, data):
    print('>>>> error <<<<< ')

@sio.event
async def message(sid, data):
    print('>>>> message <<<<< ')
    print(data)
    return 'Acknowledgement From Server'

ctx = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
ctx.load_cert_chain('server.cert', 'server.key')
web.run_app(app, host='localhost', port=2499, ssl_context=ctx)
```

#################################################################

> My client code is:

```
const io = require('socket.io-client');
var jwt = require('jsonwebtoken');

let token = jwt.sign({ username: 'git' }, '2YyZ?&qLkKus`pGV', { expiresIn: 60 * 60 });

const socket = io.connect('wss://localhost:2499', {
    forceNew: true,
    autoConnect: true,
    rejectUnauthorized: false,
    reconnection: false,
    secure: true,
    transports: ['websocket'],
    hostname: 'localhost',
    port: 2499,
    upgrade: false,
    query: {
        token
    },
})

socket.on("connect", (data, aaaaa) => {
    console.log("connect====>");
    console.log(data);
})

socket.on("connect_error", (data) => {
    console.log("connect_error====>");
    console.log((data));
})

socket.on("message", (data) => {
    console.log("message===>", data);
})

socket.on("error", (data) => {
    console.log("Error===>", data);
})

socket.on("disconnect", (data) => {
    console.log("disconnect===>", data);
});
```

What's Wrong here.Why i am not getting same error on client side?
How can i achieve desired output?

Thanks in advance

**Comment by miguelgrinberg** [comment]
What is the output from your client in those two error situations?

**Comment by harshkoralwala** [comment]
Hello @miguelgrinberg,

Thank you for Reply.

Below is my case wise error received information

################################
**Case 1**: invalid Url=>e.g wss://xyz.abc
Client side error received 
`
{ Error: websocket error
    at WS.Transport.onError (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/lib/transport.js:67:13)
    at WebSocket.ws.onerror (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/lib/transports/websocket.js:157:10)
    at WebSocket.onError (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/node_modules/ws/lib/event-target.js:128:16)
    at WebSocket.emit (events.js:198:13)
    at ClientRequest.req.on (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/node_modules/ws/lib/websocket.js:554:10)
    at ClientRequest.emit (events.js:198:13)
    at TLSSocket.socketErrorListener (_http_client.js:392:9)
    at TLSSocket.emit (events.js:198:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
  type: 'TransportError',
  description:
   ErrorEvent {
     target:
      WebSocket {
        _events: [Object],
        _eventsCount: 4,
        _maxListeners: undefined,
        readyState: 2,
        protocol: '',
        _binaryType: 'arraybuffer',
        _closeFrameReceived: false,
        _closeFrameSent: false,
        _closeMessage: '',
        _closeTimer: null,
        _closeCode: 1006,
        _extensions: {},
        _isServer: false,
        _receiver: null,
        _sender: null,
        _socket: null,
        url:
         'wss://xyz.abc/socket.io/?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhcHBsaWNhdGlvbiI6Inp1dGgiLCJpYXQiOjE1NjYzNzkyMjYsImV4cCI6MTU2NjM4MjgyNn0.iEIanwbrOM_fiZyLSzLrdshhJ5sC-bh3W3exk4SK7JM&EIO=3&transport=websocket',
        _req: null },
     type: 'error',
     message: 'getaddrinfo ENOTFOUND xyz.abc xyz.abc:443',
     error:
      { Error: getaddrinfo ENOTFOUND xyz.abc xyz.abc:443
          at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:56:26)
        errno: 'ENOTFOUND',
        code: 'ENOTFOUND',
        syscall: 'getaddrinfo',
        hostname: 'xyz.abc',
        host: 'xyz.abc',
        port: 443 } } }`

###############################



################################
**Case 2**: Valid Url but invalid Token
  here I am raising an error with...
_` raise ConnectionRefusedError('authentication failed')`_
  

Client-side error received 

`
{ Error: websocket error
    at WS.Transport.onError (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/lib/transport.js:67:13)
    at WebSocket.ws.onerror (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/lib/transports/websocket.js:157:10)
    at WebSocket.onError (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/node_modules/ws/lib/event-target.js:128:16)
    at WebSocket.emit (events.js:198:13)
    at abortHandshake (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/node_modules/ws/lib/websocket.js:673:15)
    at ClientRequest.req.on (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/node_modules/ws/lib/websocket.js:561:5)
    at ClientRequest.emit (events.js:198:13)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:556:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:109:17)
    at TLSSocket.socketOnData (_http_client.js:442:20)
  type: 'TransportError',
  description:
   ErrorEvent {
     target:
      WebSocket {
        _events: [Object],
        _eventsCount: 4,
        _maxListeners: undefined,
        readyState: 2,
        protocol: '',
        _binaryType: 'arraybuffer',
        _closeFrameReceived: false,
        _closeFrameSent: false,
        _closeMessage: '',
        _closeTimer: null,
        _closeCode: 1006,
        _extensions: {},
        _isServer: false,
        _receiver: null,
        _sender: null,
        _socket: null,
        url:
         'wss://localhost:2499/socket.io/?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhcHBsaWNhdGlvbiI6Inp1dGgiLCJpYXQiOjE1NjYzNzkzNTUsImV4cCI6MTU2NjM4Mjk1NX0.ozCh63lbHZ8StcbNrONF_bPZyz64c40YJSEaJOOJ-gc&EIO=3&transport=websocket',
        _req: [ClientRequest] },
     type: 'error',
     message: 'Unexpected server response: 401',
     error:
      Error: Unexpected server response: 401
          at ClientRequest.req.on (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/node_modules/ws/lib/websocket.js:561:5)
          at ClientRequest.emit (events.js:198:13)
          at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:556:21)
          at HTTPParser.parserOnHeadersComplete (_http_common.js:109:17)
          at TLSSocket.socketOnData (_http_client.js:442:20)
          at TLSSocket.emit (events.js:198:13)
          at addChunk (_stream_readable.js:288:12)
          at readableAddChunk (_stream_readable.js:269:11)
          at TLSSocket.Readable.push (_stream_readable.js:224:10)
          at TLSWrap.onStreamRead (internal/stream_base_commons.js:94:17) } }`

I hope this error log helps you!

Thank you

**Comment by miguelgrinberg** [comment]
The first case does not reach the server, since you are using the wrong URL, so for that one the server has no participation.

In the second case, the server side connect handler runs and raises the `ConnectionRefusedError` exception. This should have triggered an ERROR packet to be sent to the client, which I believe the JS client will pass on to your `connect_error` event handler. Have you checked if this handler is invoked? You may want to enable logging for the server and the client to see if the `ERROR` packet was sent and received. After the ERROR packet is sent, the server returns a 401 response, which is obviously unexpected by the client, and it means that the connection was not accepted due to invalid authentication.

**Comment by harshkoralwala** [comment]
Hello @miguelgrinberg,

Thank you again for the quick reply.

Case 1 of invalid URL is out of the discussion.

Yes **connect_error** handler is executed in client side(nodeJs)
you can see my first line in that handler is  ` console.log("connect_error====>");`

So my whole console after error form top looks like...

`connect_error====>
{ Error: websocket error
    at WS.Transport.onError (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/lib/transport.js:67:13)
    at WebSocket.ws.onerror (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/lib/transports/websocket.js:157:10)
    at WebSocket.onError (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/node_modules/ws/lib/event-target.js:128:16)
    at WebSocket.emit (events.js:198:13)
    at ClientRequest.req.on (/home/harsh/Harsh.koralwala/node-websocket/node_modules/engine.io-client/node_modules/ws/lib/websocket.js:554:10)
    at ClientRequest.emit (events.js:198:13)
    at TLSSocket.socketErrorListener (_http_client.js:392:9)
    at TLSSocket.emit (events.js:198:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)...` and so on.

Also, I have confirmed the handler invocation by debugging node client. my debug pointer hits the handler on error.

I have Tested My node client with NodeJs server also, Which works fine in terms of receiving the same error thrown by node server.

Just the Python server and node Client does not work properly.

For Authentication Case:
I have tried by following error raise...
 =>  return False
 =>  raise EnvironmentError('authentication failed')
 =>  raise ConnectionAbortedError('authentication failed')
 =>  raise ValueError('authentication failed')
 
But My client side still i am receiving same error as 

> Unexpected server response: 401

 for all errors.

So as per your thought...

> 
_You may want to enable logging for the server and the client to see if the ERROR packet was sent and received. After the ERROR packet is sent, the server returns a 401 response, which is obviously unexpected by the client, and it means that the connection was not accepted due to invalid_. 

This thing is not true here because as per error type different error code must be received by client.

Is there any way that i can authenticate user first and then allow for actual socket io server connection?.e.g something like before-connect event handler /  middleware like node js code...

`const io = require('socket.io')(server);
io.use((socket, next) => {
if (isValid(token)) {
    return next();
}
  return next(new Error('authentication error'));
});
`

Thanks in advance.

**Comment by miguelgrinberg** [comment]
The errors that the WebSocket client are indeed different. In case 1 you get `getaddrinfo ENOTFOUND xyz.abc xyz.abc:443`, while in case 2 you get `Unexpected server response: 401`. The fact that they chose to raise the same exception type for both errors is outside of my control, that's not something I can change from the server side.

**Comment by harshkoralwala** [comment]
Hello @miguelgrinberg,

As an alternative solution, I have done something like this...

-For Invalid Url it trows error in "connect_error" handler, which is perfect(so "connect_error" is reserved for invalid URL).
-For Valid URL and valid credential, I have emitted a custom event from the server after the credential check...

>   await sio.emit("authentication_passed",'U r Authorised User')

So here on client-side, two handlers will be called 1: "Connect" and 2. "authentication_passed"
here I am ignoring "connect" event.

So in handler "authentication_passed", I considered it a valid URL + valid user.

-For Valid URL but invalid credential, I have emitted custom event like this from the server after credential check

> await sio.emit("authentication_failed",'U r NOT Authorised User!!!')

So client-side again two custom handlers are called 1: "Connect" and 2. "authentication_failed".

Here I am ignoring the connect event. I manually disconnect the client from the server from the client-side. So no further communication possible with the same clientId.

So my  "authentication_failed" handler looks like...

`socket.on("authentication_failed", (data) => {
    console.log("authentication_Failed====>", data);
    socket.close();
})
`

My server-side code looks like...

`
async def connect(sid, environ):
    u = parse_qs(environ['QUERY_STRING'])
    data = jwt.decode(u['token'][0], '2JyZ?&qLkKLs`wGV')
    if(data['some_property'] == 'somevalue'):
       await sio.emit("authentication_passed",'U r Authorised User')
    else:
        await sio.emit("authentication_failed",'U r NOT Authorised User!!!')
`
So this way i have handled both scenarios...

Thanks
